src/rpc/hooks.zig

```diff
@@ -0,0 +1,128 @@
+            };
+        }
+
+        fn init(allocator: std.mem.Allocator, value: anytype) !*Typed(@TypeOf(value)) {
+            const TypedRef = Typed(@TypeOf(value));
+            const typed = try allocator.create(TypedRef);
+            typed.* = .{
+                .value = value,
+                .ctx_ref = .{
+                    .ref_count = 1,
+                    .free = &(struct {
+                        fn free(ctx_ref: *ContextRef, _allocator: std.mem.Allocator) void {
+                            const t: *TypedRef = @alignCast(@fieldParentPtr("ctx_ref", ctx_ref));
+                            _allocator.destroy(t);
+                        }
+                    }.free),
+                },
+            };
+            return typed;
+        }
+
+        fn inc(self: *ContextRef) *ContextRef {
+            self.ref_count += 1;
+            return self;
+        }
+
+        fn dec(self: *ContextRef, allocator: std.mem.Allocator) void {
+            self.ref_count -= 1;
+            if (self.ref_count == 0) self.free(self, allocator);
+        }
+    };
+
+    pub fn deinit(self: Hooks, allocator: std.mem.Allocator) void {
+        var it = self.map.bits.iterator(.{});
+        while (it.next()) |i| {
+            self.map.values[i].ctx_ref.dec(allocator);
+        }
+    }
+
+    pub fn set(
+        self: *Hooks,
+        allocator: std.mem.Allocator,
+        context: anytype,
+    ) !void {
+        const cref = try ContextRef.init(allocator, context);
+        defer cref.ctx_ref.dec(allocator);
+
+        const Context = @TypeOf(context);
+        inline for (comptime std.meta.declarations(Context)) |decl| {
+            const method = comptime for (std.meta.fieldNames(Method)) |method_name| {
+                if (std.mem.eql(u8, method_name, decl.name)) {
+                    break @field(Method, method_name);
+                }
+            } else @compileError("No RPC method named: " ++ decl.name);
+
+            const callback = @field(Context, decl.name);
+            if (self.map.contains(method)) {
+                return error.MethodAlreadyImplemented;
+            }
+
+            const CRef = @TypeOf(cref.*);
+            self.map.put(method, .{
+                .ctx_ref = cref.ctx_ref.inc(), // keeps the cref alive.
+                .callback = &(struct {
+                    fn impl(
+                        _allocator: std.mem.Allocator,
+                        ctx_ref: *ContextRef,
+                        args: sig.rpc.methods.Request(method),
+                    ) sig.rpc.methods.Result(method) {
+                        const _cref: *CRef = @alignCast(@fieldParentPtr("ctx_ref", ctx_ref));
+                        if (@call(.auto, callback, .{ _cref.value, _allocator, args })) |response| {
+                            return .{ .ok = response };
+                        } else |err| {
+                            return .{ .err = .{
+                                .code = @enumFromInt(@intFromError(err)),
```

Comment:
This conversion doesn't seem sound? Just the idea of returning random based-on-compilation-unit error codes as an RPC error makes me a bit uneasy, but if we do there should at least be some minimal conversion layer to make sure it doesn't happen to coincide with any of the pre-defined values in `ErrorCode`.

Reply:
JSON-RPC 2.0 [spec](https://www.jsonrpc.org/specification#error_object) only holds `-32_000`+ values noted in `sig.rpc.response.ErrorCode`, which is outside the range of numbers returned by `@intFromError()` which is a u16 or some unsigned/positive integer.

The error conditions for Solana RPC calls are marked as "unstable" (see [getHealth](https://solana.com/docs/rpc/http/gethealth)'s "Unhealthy" tab docs), so currently implementation-defined. The goal of `intFromError` casts was to make it easier for registered hook functions to return "some error or Response" instead of fiddling with what exactly to return on RPC error.

Reply:
Would be good to have a comment explaining exactly that on this bit of code.